docs/design/interceptors.md

```diff
@@ -0,0 +1,691 @@
+
+# Abstract
+
+This document covers the design and implementation of the API for extending how
+the SDK executes an operation. This extension is referred to as an *interceptor*. 
+An interceptor allows a customer to hook into specific stages of execution within 
+an SDK such as serialization or signing. It allows extending those stages by modifying 
+their inputs or outputs. Interceptors are registered during code generation or runtime
+through client configuration or runtime plugins.
+
+# Design
+
+## Terminology 
+
+* **Execution** - one end-to-end invocation against an SDK client
+
+* **Attempt** - a single attempt at performing an execution, executions may be retired multiple times 
+based on the client's retry strategy
+
+* **Hook** - a specific point during the execution of an operation that an interceptor can 
+be notified about. Hooks are either **read-only** or **read/write**. **Read-only** hooks
+allow an interceptor to read the operation input, protocol request, protocol response,
+and/or the operation output. **Read/write** hooks allow an interceptor to modify
+one of these entities. 
+
+
+## Client Interface
+
+The set of hooks is defined in the interface below with descriptions of when they are executed, what information is available,
+as well as how errors are handled.
+
+Each interceptor hook can:
+• Read the messages generated so far during in the execution
+• Save information to a context object that is available across all hooks
+• (Write Hooks Only) Modify the input message, transport request message, transport response message or output 
+
+
+```kotlin
+
+/**
+ * An interceptor allows injecting code into the request execution pipeline of a generated SDK client.
+ *
+ * Terminology:
+ * * `execution` - one end-to-end invocation against an SDK client
+ * * `attempt` - a single attempt at performing an execution, executions may be retired multiple times
+ * based on the client's retry strategy.
+ * * `hook` - a single method on the interceptor allowing injection of code into a specific part of the execution
+ * pipeline. Hooks are either "read-only" hooks, which make it possible to read in-flight request or response messages,
+ * or `read/write` hooks, which make it possible to modify in-flight request or response messages. Read only hooks
+ * **MUST** not modify state even if it is possible to do so (it is not always possible or performant to provide an
+ * immutable view of every type).
+ */
+public interface Interceptor<
+    Input,
+    Output,
+    ProtocolRequest,
+    ProtocolResponse,
+    > {
+
+    /**
+     * A hook called at the start of an execution, before the SDK does anything else.
+     *
+     * **When**: This will ALWAYS be called once per execution. The duration between
+     * invocation of this hook and [readAfterExecution] is very close to full duration of the execution
+     *
+     * **Available Information**: [RequestInterceptorContext.request] is always available.
```

Comment:
Ya I can take a look at removing that, it is reflected in the types. It may still be useful in some cases (e.g. `modifyBeforeCompletion`) as it explains why some of the types could be `null`. For the rest it can probably be left off.